map.go

```diff
@@ -0,0 +1,144 @@
+package validation
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"reflect"
+)
+
+var (
+	// ErrNotMap is the error that the value being validated is not a map.
+	ErrNotMap = errors.New("only a map can be validated")
+
+	// ErrKeyWrongType is the error returned in case of an incorrect key type.
+	ErrKeyWrongType = NewError("validation_key_wrong_type", "key not the correct type")
+
+	// ErrKeyNotFound is the error returned in case of a missing key.
+	ErrKeyNotFound = NewError("validation_key_missing", "key not found")
```

Comment:
Could you rename it to `ErrKeyMissing`? This is because `ErrKeyNotFound` sounds like we are looking for a key and it's not found. Also the error message can be updated to `required key is missing`

Reply:
Done. Also tidied up commits and rebased on `master` for good measure.